vllm/distributed/device_communicators/all2all.py

```diff
@@ -0,0 +1,93 @@
+# SPDX-License-Identifier: Apache-2.0
+import torch
+
+from vllm.forward_context import get_forward_context
+
+
+class All2AllBase:
+
+    def __init__(self, cpu_group, model):
+        self.cpu_group = cpu_group
+
+        # compute some common properties
+        from vllm.distributed.parallel_state import (get_dp_group,
+                                                     get_ep_group,
+                                                     get_tp_group,
+                                                     in_the_same_node_as)
+
+        # all2all lives in ep group, which is merged from dp and tp group
+        self.dp_group = get_dp_group()
+        self.tp_group = get_tp_group()
+        self.ep_group = get_ep_group()
+        self.dp_rank = self.dp_group.rank_in_group
+        self.dp_world_size = self.dp_group.world_size
+
+        # all2all communication often has separate implementations for
+        # intra-node and inter-node communication
+        self.intranode = in_the_same_node_as(cpu_group, source_rank=0)
+        self.internode = not self.intranode
+
+    def dispatch(self, hidden_states: torch.Tensor,
+                 router_logits: torch.Tensor):
+        raise NotImplementedError
+
+    def combine(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        raise NotImplementedError
```

Comment:
Is there a function signature for dispatch and combine that will work across all kernels?
* The naive multicast implementation
* pplx-kernels
* DeepEP

I think it would to look like:
```
def dispatch(self, 
             hidden_states: torch.Tensor,
             hs_scales: torch.Tensor, # Quantized scales for hidden_states
             router_logits: torch.Tensor,  
             topk_weights: torch.Tensor,  
             topk_ids: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor] # out_hidden_states and out_scales
```
(There are other parameters to the pplx-kernels that are needed but I think they can be encapsulated in a PPLXAll2All wrapper class)

@varun-sundar-rabindranath @bnellnm WDYT?

Reply:
Another consideration here is that the output format of the different A2A implementations will be different.

* For naive All2All, the output shape would be `[total_tokens_across_dp, hidden_size]`
* For pplx All2All, it will be `[num_experts_per_rank, max_tokens_per_expert, hidden_size]`, with padding along axis 1
* For DeepEP I am not sure what the output format will look like

Reply:
for a generic dispatch/combine interface for different backends (pplx, naive, deepep), I think we can have:
```python
def dispatch(self, tensors: List[torch.Tensor]) -> List[torch.Tensor]:
def combine(self, tensors: List[torch.Tensor]) -> List[torch.Tensor]:
```
the backend can decide what does each tensor mean, and the moe layer can use assertion to make sure the list contains data it needs.

Reply:
What's the reason for having the base class in this case? It seems that the caller will have to know exactly the implementation it's using

Reply:
It's mainly to manage the lifecycle of the communication library, and separate the implementations.

it is true that the dispatch / combine apis are quite different, but what matters in this pr is to unify the construction, backend selection, and deconstruction. as a byproduct, we need to have a generic dispatch / combine function.

Reply:
I agree with @tlrmchlsmth that it would be good if these interfaces matched the pplx AllToAll interfaces.

Reply:
feel free to change the interface in the moe modularization pr!